// @magicblock-labs/ephemeral-rollups-sdk/lib/instructions/ephemeral-spl-token-program/ephemeralAta.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.decodeEphemeralAta = decodeEphemeralAta;
exports.encodeEphemeralAta = encodeEphemeralAta;
exports.decodeGlobalVault = decodeGlobalVault;
exports.encodeGlobalVault = encodeGlobalVault;
exports.deriveEphemeralAta = deriveEphemeralAta;
exports.deriveVault = deriveVault;
exports.deriveRentPda = deriveRentPda;
exports.deriveVaultAta = deriveVaultAta;
exports.deriveShuttleEphemeralAta = deriveShuttleEphemeralAta;
exports.deriveShuttleAta = deriveShuttleAta;
exports.deriveShuttleWalletAta = deriveShuttleWalletAta;
exports.initEphemeralAtaIx = initEphemeralAtaIx;
exports.initVaultAtaIx = initVaultAtaIx;
exports.initVaultIx = initVaultIx;
exports.initRentPdaIx = initRentPdaIx;
exports.transferToVaultIx = transferToVaultIx;
exports.depositSplTokensIx = depositSplTokensIx;
exports.delegateEphemeralAtaIx = delegateEphemeralAtaIx;
exports.initShuttleEphemeralAtaIx = initShuttleEphemeralAtaIx;
exports.delegateShuttleEphemeralAtaIx = delegateShuttleEphemeralAtaIx;
exports.setupAndDelegateShuttleEphemeralAtaWithMergeIx = setupAndDelegateShuttleEphemeralAtaWithMergeIx;
exports.depositAndDelegateShuttleEphemeralAtaWithMergeAndPrivateTransferIx = depositAndDelegateShuttleEphemeralAtaWithMergeAndPrivateTransferIx;
exports.withdrawThroughDelegatedShuttleWithMergeIx = withdrawThroughDelegatedShuttleWithMergeIx;
exports.mergeShuttleIntoAtaIx = mergeShuttleIntoAtaIx;
exports.undelegateAndCloseShuttleEphemeralAtaIx = undelegateAndCloseShuttleEphemeralAtaIx;
exports.withdrawSplIx = withdrawSplIx;
exports.undelegateIx = undelegateIx;
exports.createEataPermissionIx = createEataPermissionIx;
exports.resetEataPermissionIx = resetEataPermissionIx;
exports.delegateEataPermissionIx = delegateEataPermissionIx;
exports.undelegateEataPermissionIx = undelegateEataPermissionIx;
exports.delegateSpl = delegateSpl;
exports.delegateSplWithPrivateTransfer = delegateSplWithPrivateTransfer;
exports.transferSpl = transferSpl;
exports.withdrawSpl = withdrawSpl;
const web3_js_1 = require("@solana/web3.js");
const constants_js_1 = require("../../constants.js");
const pda_js_1 = require("../../pda.js");
const transferQueue_js_1 = require("./transferQueue.js");
const TOKEN_PROGRAM_ID = new web3_js_1.PublicKey("TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA");
const ASSOCIATED_TOKEN_PROGRAM_ID = new web3_js_1.PublicKey("ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL");
function getAssociatedTokenAddressSync(mint, owner, allowOwnerOffCurve = false, programId = TOKEN_PROGRAM_ID, associatedTokenProgramId = ASSOCIATED_TOKEN_PROGRAM_ID) {
    if (!allowOwnerOffCurve && !web3_js_1.PublicKey.isOnCurve(owner)) {
        throw new Error("Owner public key is off-curve");
    }
    const [ata] = web3_js_1.PublicKey.findProgramAddressSync([owner.toBuffer(), programId.toBuffer(), mint.toBuffer()], associatedTokenProgramId);
    return ata;
}
function createAssociatedTokenAccountIdempotentInstruction(payer, associatedToken, owner, mint, programId = TOKEN_PROGRAM_ID, associatedTokenProgramId = ASSOCIATED_TOKEN_PROGRAM_ID) {
    const data = Buffer.from([1]);
    return new web3_js_1.TransactionInstruction({
        programId: associatedTokenProgramId,
        keys: [
            { pubkey: payer, isSigner: true, isWritable: true },
            { pubkey: associatedToken, isSigner: false, isWritable: true },
            { pubkey: owner, isSigner: false, isWritable: false },
            { pubkey: mint, isSigner: false, isWritable: false },
            { pubkey: web3_js_1.SystemProgram.programId, isSigner: false, isWritable: false },
            { pubkey: programId, isSigner: false, isWritable: false },
        ],
        data,
    });
}
function createTransferInstruction(source, destination, owner, amount, multiSigners = [], programId = TOKEN_PROGRAM_ID) {
    const data = Buffer.alloc(9);
    data[0] = 3;
    data.writeBigUInt64LE(amount, 1);
    const keys = [
        { pubkey: source, isSigner: false, isWritable: true },
        { pubkey: destination, isSigner: false, isWritable: true },
    ];
    if (multiSigners.length === 0) {
        keys.push({ pubkey: owner, isSigner: true, isWritable: false });
    }
    else {
        keys.push({ pubkey: owner, isSigner: false, isWritable: false });
        for (const signer of multiSigners) {
            keys.push({ pubkey: signer, isSigner: true, isWritable: false });
        }
    }
    return new web3_js_1.TransactionInstruction({
        programId,
        keys,
        data,
    });
}
function decodeEphemeralAta(info) {
    if (info.data.length < 72) {
        throw new Error("Invalid EphemeralAta account data length");
    }
    const owner = new web3_js_1.PublicKey(info.data.subarray(0, 32));
    const mint = new web3_js_1.PublicKey(info.data.subarray(32, 64));
    const amount = BigInt(info.data.readBigUInt64LE(64));
    return {
        owner,
        mint,
        amount,
    };
}
function encodeEphemeralAta(eata) {
    const buffer = Buffer.alloc(72);
    buffer.set(eata.owner.toBytes(), 0);
    buffer.set(eata.mint.toBytes(), 32);
    buffer.writeBigUInt64LE(eata.amount, 64);
    return buffer;
}
function decodeGlobalVault(info) {
    if (info.data.length < 32) {
        throw new Error("Invalid GlobalVault account data length");
    }
    const mint = new web3_js_1.PublicKey(info.data.subarray(0, 32));
    return { mint };
}
function encodeGlobalVault(vault) {
    const buffer = Buffer.alloc(32);
    buffer.set(vault.mint.toBytes(), 0);
    return buffer;
}
function deriveEphemeralAta(owner, mint) {
    return web3_js_1.PublicKey.findProgramAddressSync([owner.toBuffer(), mint.toBuffer()], constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID);
}
function deriveVault(mint) {
    return web3_js_1.PublicKey.findProgramAddressSync([mint.toBuffer()], constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID);
}
function deriveRentPda() {
    return web3_js_1.PublicKey.findProgramAddressSync([Buffer.from("rent")], constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID);
}
function deriveVaultAta(mint, vault) {
    return getAssociatedTokenAddressSync(mint, vault, true);
}
function deriveShuttleEphemeralAta(owner, mint, shuttleId) {
    if (!Number.isInteger(shuttleId) ||
        shuttleId < 0 ||
        shuttleId > 4294967295) {
        throw new Error("shuttleId must fit in u32");
    }
    const shuttleIdSeed = Buffer.alloc(4);
    shuttleIdSeed.writeUInt32LE(shuttleId, 0);
    return web3_js_1.PublicKey.findProgramAddressSync([owner.toBuffer(), mint.toBuffer(), shuttleIdSeed], constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID);
}
function deriveShuttleAta(shuttleEphemeralAta, mint) {
    return web3_js_1.PublicKey.findProgramAddressSync([shuttleEphemeralAta.toBuffer(), mint.toBuffer()], constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID);
}
function deriveShuttleWalletAta(mint, shuttleEphemeralAta) {
    return getAssociatedTokenAddressSync(mint, shuttleEphemeralAta, true);
}
function initEphemeralAtaIx(ephemeralAta, owner, mint, payer, bump) {
    return new web3_js_1.TransactionInstruction({
        programId: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
        keys: [
            { pubkey: ephemeralAta, isSigner: false, isWritable: true },
            { pubkey: payer, isSigner: true, isWritable: true },
            { pubkey: owner, isSigner: false, isWritable: false },
            { pubkey: mint, isSigner: false, isWritable: false },
            { pubkey: web3_js_1.SystemProgram.programId, isSigner: false, isWritable: false },
        ],
        data: Buffer.from([0, bump]),
    });
}
function initVaultAtaIx(payer, vaultAta, vault, mint) {
    return createAssociatedTokenAccountIdempotentInstruction(payer, vaultAta, vault, mint);
}
function initVaultIx(vault, mint, payer, bump) {
    const [vaultEphemeralAta] = deriveEphemeralAta(vault, mint);
    const vaultAta = deriveVaultAta(mint, vault);
    return new web3_js_1.TransactionInstruction({
        programId: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
        keys: [
            { pubkey: vault, isSigner: false, isWritable: true },
            { pubkey: payer, isSigner: true, isWritable: true },
            { pubkey: mint, isSigner: false, isWritable: false },
            { pubkey: vaultEphemeralAta, isSigner: false, isWritable: true },
            { pubkey: vaultAta, isSigner: false, isWritable: true },
            { pubkey: TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
            {
                pubkey: ASSOCIATED_TOKEN_PROGRAM_ID,
                isSigner: false,
                isWritable: false,
            },
            { pubkey: web3_js_1.SystemProgram.programId, isSigner: false, isWritable: false },
        ],
        data: Buffer.from([1, bump]),
    });
}
function initRentPdaIx(payer, rentPda) {
    return new web3_js_1.TransactionInstruction({
        programId: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
        keys: [
            { pubkey: payer, isSigner: true, isWritable: true },
            { pubkey: rentPda, isSigner: false, isWritable: true },
            { pubkey: web3_js_1.SystemProgram.programId, isSigner: false, isWritable: false },
        ],
        data: Buffer.from([23]),
    });
}
function transferToVaultIx(ephemeralAta, vault, mint, sourceAta, vaultAta, owner, amount) {
    return new web3_js_1.TransactionInstruction({
        programId: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
        keys: [
            { pubkey: ephemeralAta, isSigner: false, isWritable: true },
            { pubkey: vault, isSigner: false, isWritable: false },
            { pubkey: mint, isSigner: false, isWritable: false },
            { pubkey: sourceAta, isSigner: false, isWritable: true },
            { pubkey: vaultAta, isSigner: false, isWritable: true },
            { pubkey: owner, isSigner: true, isWritable: false },
            { pubkey: TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
        ],
        data: encodeAmountInstructionData(2, amount),
    });
}
function depositSplTokensIx(ephemeralAta, vault, mint, sourceAta, vaultAta, owner, amount) {
    return transferToVaultIx(ephemeralAta, vault, mint, sourceAta, vaultAta, owner, amount);
}
function delegateEphemeralAtaIx(payer, ephemeralAta, bump, validator) {
    const data = validator
        ? Buffer.concat([Buffer.from([4, bump]), validator.toBuffer()])
        : Buffer.from([4, bump]);
    return new web3_js_1.TransactionInstruction({
        programId: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
        keys: [
            { pubkey: payer, isSigner: true, isWritable: true },
            { pubkey: ephemeralAta, isSigner: false, isWritable: true },
            {
                pubkey: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
                isSigner: false,
                isWritable: false,
            },
            {
                pubkey: (0, pda_js_1.delegateBufferPdaFromDelegatedAccountAndOwnerProgram)(ephemeralAta, constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID),
                isSigner: false,
                isWritable: true,
            },
            {
                pubkey: (0, pda_js_1.delegationRecordPdaFromDelegatedAccount)(ephemeralAta),
                isSigner: false,
                isWritable: true,
            },
            {
                pubkey: (0, pda_js_1.delegationMetadataPdaFromDelegatedAccount)(ephemeralAta),
                isSigner: false,
                isWritable: true,
            },
            { pubkey: constants_js_1.DELEGATION_PROGRAM_ID, isSigner: false, isWritable: false },
            { pubkey: web3_js_1.SystemProgram.programId, isSigner: false, isWritable: false },
        ],
        data,
    });
}
function initShuttleEphemeralAtaIx(payer, shuttleEphemeralAta, shuttleAta, shuttleWalletAta, owner, mint, shuttleId, bump) {
    if (!Number.isInteger(shuttleId) ||
        shuttleId < 0 ||
        shuttleId > 4294967295) {
        throw new Error("shuttleId must fit in u32");
    }
    const data = Buffer.alloc(6);
    data[0] = 11;
    data.writeUInt32LE(shuttleId, 1);
    data[5] = bump;
    return new web3_js_1.TransactionInstruction({
        programId: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
        keys: [
            { pubkey: payer, isSigner: true, isWritable: true },
            { pubkey: shuttleEphemeralAta, isSigner: false, isWritable: true },
            { pubkey: shuttleAta, isSigner: false, isWritable: true },
            { pubkey: shuttleWalletAta, isSigner: false, isWritable: true },
            { pubkey: owner, isSigner: false, isWritable: false },
            { pubkey: mint, isSigner: false, isWritable: false },
            { pubkey: TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
            {
                pubkey: ASSOCIATED_TOKEN_PROGRAM_ID,
                isSigner: false,
                isWritable: false,
            },
            { pubkey: web3_js_1.SystemProgram.programId, isSigner: false, isWritable: false },
        ],
        data,
    });
}
function delegateShuttleEphemeralAtaIx(payer, shuttleEphemeralAta, shuttleAta, bump, validator) {
    const data = validator
        ? Buffer.concat([Buffer.from([13, bump]), validator.toBuffer()])
        : Buffer.from([13, bump]);
    return new web3_js_1.TransactionInstruction({
        programId: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
        keys: [
            { pubkey: payer, isSigner: true, isWritable: true },
            { pubkey: shuttleEphemeralAta, isSigner: false, isWritable: false },
            { pubkey: shuttleAta, isSigner: false, isWritable: true },
            {
                pubkey: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
                isSigner: false,
                isWritable: false,
            },
            {
                pubkey: (0, pda_js_1.delegateBufferPdaFromDelegatedAccountAndOwnerProgram)(shuttleAta, constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID),
                isSigner: false,
                isWritable: true,
            },
            {
                pubkey: (0, pda_js_1.delegationRecordPdaFromDelegatedAccount)(shuttleAta),
                isSigner: false,
                isWritable: true,
            },
            {
                pubkey: (0, pda_js_1.delegationMetadataPdaFromDelegatedAccount)(shuttleAta),
                isSigner: false,
                isWritable: true,
            },
            { pubkey: constants_js_1.DELEGATION_PROGRAM_ID, isSigner: false, isWritable: false },
            { pubkey: web3_js_1.SystemProgram.programId, isSigner: false, isWritable: false },
        ],
        data,
    });
}
function setupAndDelegateShuttleEphemeralAtaWithMergeIx(payer, shuttleEphemeralAta, shuttleAta, owner, sourceAta, destinationAta, shuttleWalletAta, mint, shuttleId, bump, amount, validator) {
    if (!Number.isInteger(shuttleId) ||
        shuttleId < 0 ||
        shuttleId > 4294967295) {
        throw new Error("shuttleId must fit in u32");
    }
    const [rentPda] = deriveRentPda();
    const [vault] = deriveVault(mint);
    const vaultAta = deriveVaultAta(mint, vault);
    const data = validator ? Buffer.alloc(46) : Buffer.alloc(14);
    data[0] = 24;
    data.writeUInt32LE(shuttleId, 1);
    data[5] = bump;
    data.writeBigUInt64LE(amount, 6);
    if (validator) {
        validator.toBuffer().copy(data, 14);
    }
    return new web3_js_1.TransactionInstruction({
        programId: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
        keys: [
            { pubkey: payer, isSigner: true, isWritable: true },
            { pubkey: rentPda, isSigner: false, isWritable: true },
            { pubkey: shuttleEphemeralAta, isSigner: false, isWritable: true },
            { pubkey: shuttleAta, isSigner: false, isWritable: true },
            { pubkey: shuttleWalletAta, isSigner: false, isWritable: true },
            { pubkey: owner, isSigner: true, isWritable: false },
            {
                pubkey: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
                isSigner: false,
                isWritable: false,
            },
            {
                pubkey: (0, pda_js_1.delegateBufferPdaFromDelegatedAccountAndOwnerProgram)(shuttleAta, constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID),
                isSigner: false,
                isWritable: true,
            },
            {
                pubkey: (0, pda_js_1.delegationRecordPdaFromDelegatedAccount)(shuttleAta),
                isSigner: false,
                isWritable: true,
            },
            {
                pubkey: (0, pda_js_1.delegationMetadataPdaFromDelegatedAccount)(shuttleAta),
                isSigner: false,
                isWritable: true,
            },
            { pubkey: constants_js_1.DELEGATION_PROGRAM_ID, isSigner: false, isWritable: false },
            {
                pubkey: ASSOCIATED_TOKEN_PROGRAM_ID,
                isSigner: false,
                isWritable: false,
            },
            { pubkey: web3_js_1.SystemProgram.programId, isSigner: false, isWritable: false },
            { pubkey: destinationAta, isSigner: false, isWritable: true },
            { pubkey: mint, isSigner: false, isWritable: false },
            { pubkey: TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
            { pubkey: vault, isSigner: false, isWritable: false },
            { pubkey: sourceAta, isSigner: false, isWritable: true },
            { pubkey: vaultAta, isSigner: false, isWritable: true },
        ],
        data,
    });
}
function depositAndDelegateShuttleEphemeralAtaWithMergeAndPrivateTransferIx(payer, shuttleEphemeralAta, shuttleAta, owner, sourceAta, destinationAta, shuttleWalletAta, mint, shuttleId, bump, amount, minDelayMs, maxDelayMs, split, validator) {
    if (!Number.isInteger(shuttleId) ||
        shuttleId < 0 ||
        shuttleId > 4294967295) {
        throw new Error("shuttleId must fit in u32");
    }
    if (!Number.isInteger(split) || split <= 0 || split > 4294967295) {
        throw new Error("split must fit in u32");
    }
    if (amount < 0n || minDelayMs < 0n || maxDelayMs < 0n) {
        throw new Error("amount and delays must be non-negative");
    }
    if (maxDelayMs < minDelayMs) {
        throw new Error("maxDelayMs must be greater than or equal to minDelayMs");
    }
    const [rentPda] = deriveRentPda();
    const [vault] = deriveVault(mint);
    const vaultAta = deriveVaultAta(mint, vault);
    const [queue] = (0, transferQueue_js_1.deriveTransferQueue)(mint);
    const data = validator ? Buffer.alloc(66) : Buffer.alloc(34);
    data[0] = 25;
    data.writeUInt32LE(shuttleId, 1);
    data[5] = bump;
    data.writeBigUInt64LE(amount, 6);
    data.writeBigUInt64LE(minDelayMs, 14);
    data.writeBigUInt64LE(maxDelayMs, 22);
    data.writeUInt32LE(split, 30);
    if (validator) {
        validator.toBuffer().copy(data, 34);
    }
    return new web3_js_1.TransactionInstruction({
        programId: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
        keys: [
            { pubkey: payer, isSigner: true, isWritable: true },
            { pubkey: rentPda, isSigner: false, isWritable: true },
            { pubkey: shuttleEphemeralAta, isSigner: false, isWritable: true },
            { pubkey: shuttleAta, isSigner: false, isWritable: true },
            { pubkey: shuttleWalletAta, isSigner: false, isWritable: true },
            { pubkey: owner, isSigner: true, isWritable: false },
            {
                pubkey: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
                isSigner: false,
                isWritable: false,
            },
            {
                pubkey: (0, pda_js_1.delegateBufferPdaFromDelegatedAccountAndOwnerProgram)(shuttleAta, constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID),
                isSigner: false,
                isWritable: true,
            },
            {
                pubkey: (0, pda_js_1.delegationRecordPdaFromDelegatedAccount)(shuttleAta),
                isSigner: false,
                isWritable: true,
            },
            {
                pubkey: (0, pda_js_1.delegationMetadataPdaFromDelegatedAccount)(shuttleAta),
                isSigner: false,
                isWritable: true,
            },
            { pubkey: constants_js_1.DELEGATION_PROGRAM_ID, isSigner: false, isWritable: false },
            {
                pubkey: ASSOCIATED_TOKEN_PROGRAM_ID,
                isSigner: false,
                isWritable: false,
            },
            { pubkey: web3_js_1.SystemProgram.programId, isSigner: false, isWritable: false },
            { pubkey: destinationAta, isSigner: false, isWritable: true },
            { pubkey: mint, isSigner: false, isWritable: false },
            { pubkey: TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
            { pubkey: vault, isSigner: false, isWritable: false },
            { pubkey: sourceAta, isSigner: false, isWritable: true },
            { pubkey: vaultAta, isSigner: false, isWritable: true },
            { pubkey: queue, isSigner: false, isWritable: true },
        ],
        data,
    });
}
function withdrawThroughDelegatedShuttleWithMergeIx(payer, shuttleEphemeralAta, shuttleAta, owner, ownerAta, shuttleWalletAta, mint, shuttleId, bump, amount, validator) {
    if (!Number.isInteger(shuttleId) ||
        shuttleId < 0 ||
        shuttleId > 4294967295) {
        throw new Error("shuttleId must fit in u32");
    }
    if (amount < 0n) {
        throw new Error("amount must be non-negative");
    }
    const [rentPda] = deriveRentPda();
    const data = validator ? Buffer.alloc(46) : Buffer.alloc(14);
    data[0] = 26;
    data.writeUInt32LE(shuttleId, 1);
    data[5] = bump;
    data.writeBigUInt64LE(amount, 6);
    if (validator) {
        validator.toBuffer().copy(data, 14);
    }
    return new web3_js_1.TransactionInstruction({
        programId: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
        keys: [
            { pubkey: payer, isSigner: true, isWritable: true },
            { pubkey: rentPda, isSigner: false, isWritable: true },
            { pubkey: shuttleEphemeralAta, isSigner: false, isWritable: true },
            { pubkey: shuttleAta, isSigner: false, isWritable: true },
            { pubkey: shuttleWalletAta, isSigner: false, isWritable: true },
            { pubkey: owner, isSigner: true, isWritable: false },
            {
                pubkey: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
                isSigner: false,
                isWritable: false,
            },
            {
                pubkey: (0, pda_js_1.delegateBufferPdaFromDelegatedAccountAndOwnerProgram)(shuttleAta, constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID),
                isSigner: false,
                isWritable: true,
            },
            {
                pubkey: (0, pda_js_1.delegationRecordPdaFromDelegatedAccount)(shuttleAta),
                isSigner: false,
                isWritable: true,
            },
            {
                pubkey: (0, pda_js_1.delegationMetadataPdaFromDelegatedAccount)(shuttleAta),
                isSigner: false,
                isWritable: true,
            },
            { pubkey: constants_js_1.DELEGATION_PROGRAM_ID, isSigner: false, isWritable: false },
            {
                pubkey: ASSOCIATED_TOKEN_PROGRAM_ID,
                isSigner: false,
                isWritable: false,
            },
            { pubkey: web3_js_1.SystemProgram.programId, isSigner: false, isWritable: false },
            { pubkey: ownerAta, isSigner: false, isWritable: true },
            { pubkey: mint, isSigner: false, isWritable: false },
            { pubkey: TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
        ],
        data,
    });
}
function mergeShuttleIntoAtaIx(owner, destinationAta, shuttleEphemeralAta, shuttleWalletAta, mint) {
    return new web3_js_1.TransactionInstruction({
        programId: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
        keys: [
            { pubkey: owner, isSigner: true, isWritable: false },
            { pubkey: destinationAta, isSigner: false, isWritable: true },
            { pubkey: shuttleEphemeralAta, isSigner: false, isWritable: false },
            { pubkey: shuttleWalletAta, isSigner: false, isWritable: true },
            { pubkey: mint, isSigner: false, isWritable: false },
            { pubkey: TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
        ],
        data: Buffer.from([15]),
    });
}
function undelegateAndCloseShuttleEphemeralAtaIx(payer, shuttleEphemeralAta, shuttleAta, shuttleWalletAta, escrowIndex) {
    const data = escrowIndex === undefined
        ? Buffer.from([14])
        : Buffer.from([14, escrowIndex]);
    return new web3_js_1.TransactionInstruction({
        programId: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
        keys: [
            { pubkey: payer, isSigner: true, isWritable: true },
            { pubkey: shuttleEphemeralAta, isSigner: false, isWritable: false },
            { pubkey: shuttleAta, isSigner: false, isWritable: false },
            { pubkey: shuttleWalletAta, isSigner: false, isWritable: true },
            { pubkey: TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
            { pubkey: constants_js_1.MAGIC_CONTEXT_ID, isSigner: false, isWritable: true },
            { pubkey: constants_js_1.MAGIC_PROGRAM_ID, isSigner: false, isWritable: false },
        ],
        data,
    });
}
function withdrawSplIx(owner, mint, amount) {
    const [ephemeralAta] = deriveEphemeralAta(owner, mint);
    const [vault, vaultBump] = deriveVault(mint);
    const vaultAta = deriveVaultAta(mint, vault);
    const userDestAta = getAssociatedTokenAddressSync(mint, owner);
    return new web3_js_1.TransactionInstruction({
        programId: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
        keys: [
            { pubkey: owner, isSigner: true, isWritable: false },
            { pubkey: ephemeralAta, isSigner: false, isWritable: true },
            { pubkey: vault, isSigner: false, isWritable: false },
            { pubkey: mint, isSigner: false, isWritable: false },
            { pubkey: vaultAta, isSigner: false, isWritable: true },
            { pubkey: userDestAta, isSigner: false, isWritable: true },
            { pubkey: TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
        ],
        data: encodeAmountInstructionData(3, amount, vaultBump),
    });
}
function undelegateIx(owner, mint) {
    const userAta = getAssociatedTokenAddressSync(mint, owner);
    const [ephemeralAta] = deriveEphemeralAta(owner, mint);
    return new web3_js_1.TransactionInstruction({
        programId: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
        keys: [
            {
                pubkey: owner,
                isSigner: true,
                isWritable: false,
            },
            {
                pubkey: userAta,
                isSigner: false,
                isWritable: true,
            },
            {
                pubkey: ephemeralAta,
                isSigner: false,
                isWritable: false,
            },
            {
                pubkey: constants_js_1.MAGIC_CONTEXT_ID,
                isSigner: false,
                isWritable: true,
            },
            {
                pubkey: constants_js_1.MAGIC_PROGRAM_ID,
                isSigner: false,
                isWritable: false,
            },
        ],
        data: Buffer.from([5]),
    });
}
function createEataPermissionIx(ephemeralAta, payer, bump, flags = 0) {
    const permission = (0, pda_js_1.permissionPdaFromAccount)(ephemeralAta);
    return new web3_js_1.TransactionInstruction({
        programId: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
        keys: [
            { pubkey: ephemeralAta, isSigner: false, isWritable: true },
            { pubkey: permission, isSigner: false, isWritable: true },
            { pubkey: payer, isSigner: true, isWritable: true },
            { pubkey: web3_js_1.SystemProgram.programId, isSigner: false, isWritable: false },
            { pubkey: constants_js_1.PERMISSION_PROGRAM_ID, isSigner: false, isWritable: false },
        ],
        data: Buffer.from([6, bump, flags]),
    });
}
function resetEataPermissionIx(ephemeralAta, payer, bump, flags = 0) {
    const permission = (0, pda_js_1.permissionPdaFromAccount)(ephemeralAta);
    return new web3_js_1.TransactionInstruction({
        programId: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
        keys: [
            { pubkey: ephemeralAta, isSigner: false, isWritable: false },
            { pubkey: permission, isSigner: false, isWritable: true },
            { pubkey: payer, isSigner: true, isWritable: false },
            { pubkey: constants_js_1.PERMISSION_PROGRAM_ID, isSigner: false, isWritable: false },
        ],
        data: Buffer.from([9, bump, flags]),
    });
}
function delegateEataPermissionIx(payer, ephemeralAta, bump, validator) {
    const permission = (0, pda_js_1.permissionPdaFromAccount)(ephemeralAta);
    return new web3_js_1.TransactionInstruction({
        programId: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
        keys: [
            { pubkey: payer, isSigner: true, isWritable: true },
            { pubkey: ephemeralAta, isSigner: false, isWritable: true },
            { pubkey: constants_js_1.PERMISSION_PROGRAM_ID, isSigner: false, isWritable: false },
            { pubkey: permission, isSigner: false, isWritable: true },
            { pubkey: web3_js_1.SystemProgram.programId, isSigner: false, isWritable: false },
            {
                pubkey: (0, pda_js_1.delegateBufferPdaFromDelegatedAccountAndOwnerProgram)(permission, constants_js_1.PERMISSION_PROGRAM_ID),
                isSigner: false,
                isWritable: true,
            },
            {
                pubkey: (0, pda_js_1.delegationRecordPdaFromDelegatedAccount)(permission),
                isSigner: false,
                isWritable: true,
            },
            {
                pubkey: (0, pda_js_1.delegationMetadataPdaFromDelegatedAccount)(permission),
                isSigner: false,
                isWritable: true,
            },
            { pubkey: constants_js_1.DELEGATION_PROGRAM_ID, isSigner: false, isWritable: false },
            { pubkey: validator, isSigner: false, isWritable: false },
        ],
        data: Buffer.from([7, bump]),
    });
}
function undelegateEataPermissionIx(owner, ephemeralAta) {
    const permission = (0, pda_js_1.permissionPdaFromAccount)(ephemeralAta);
    return new web3_js_1.TransactionInstruction({
        programId: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
        keys: [
            { pubkey: owner, isSigner: true, isWritable: false },
            { pubkey: ephemeralAta, isSigner: false, isWritable: true },
            { pubkey: permission, isSigner: false, isWritable: true },
            { pubkey: constants_js_1.PERMISSION_PROGRAM_ID, isSigner: false, isWritable: false },
            { pubkey: constants_js_1.MAGIC_PROGRAM_ID, isSigner: false, isWritable: false },
            { pubkey: constants_js_1.MAGIC_CONTEXT_ID, isSigner: false, isWritable: true },
        ],
        data: Buffer.from([8]),
    });
}
function randomShuttleId() {
    const cryptoObj = globalThis?.crypto;
    if (cryptoObj?.getRandomValues !== undefined) {
        const buf = new Uint32Array(1);
        cryptoObj.getRandomValues(buf);
        return buf[0];
    }
    return Math.floor(Math.random() * 4294967296);
}
async function buildDelegateSplInstructions(owner, mint, amount, opts) {
    const payer = opts?.payer ?? owner;
    const validator = opts?.validator;
    const initIfMissing = opts?.initIfMissing ?? true;
    const initVaultIfMissing = opts?.initVaultIfMissing ?? initIfMissing;
    const isPrivate = opts?.private ?? false;
    const instructions = [];
    const [ephemeralAta, eataBump] = deriveEphemeralAta(owner, mint);
    const [vault, vaultBump] = deriveVault(mint);
    const [vaultEphemeralAta, vaultEataBump] = deriveEphemeralAta(vault, mint);
    const vaultAta = deriveVaultAta(mint, vault);
    const ownerAta = getAssociatedTokenAddressSync(mint, owner);
    if (initIfMissing) {
        instructions.push(initEphemeralAtaIx(ephemeralAta, owner, mint, payer, eataBump));
    }
    if (initVaultIfMissing) {
        instructions.push(initVaultIx(vault, mint, payer, vaultBump), initVaultAtaIx(payer, vaultAta, vault, mint), delegateEphemeralAtaIx(payer, vaultEphemeralAta, vaultEataBump, validator));
    }
    instructions.push(transferToVaultIx(ephemeralAta, vault, mint, ownerAta, vaultAta, owner, amount));
    if (isPrivate) {
        instructions.push(createEataPermissionIx(ephemeralAta, payer, eataBump));
    }
    instructions.push(delegateEphemeralAtaIx(payer, ephemeralAta, eataBump, validator));
    return instructions;
}
async function buildIdempotentDelegateSplInstructions(owner, mint, amount, opts) {
    const payer = opts?.payer ?? owner;
    const validator = opts?.validator;
    const initIfMissing = opts?.initIfMissing ?? true;
    const initVaultIfMissing = opts?.initVaultIfMissing ?? false;
    const initAtasIfMissing = opts?.initAtasIfMissing ?? false;
    const isPrivate = opts?.private ?? false;
    const shuttleId = opts?.shuttleId ?? randomShuttleId();
    const instructions = [];
    const [ephemeralAta, eataBump] = deriveEphemeralAta(owner, mint);
    const [vault, vaultBump] = deriveVault(mint);
    const [vaultEphemeralAta, vaultEataBump] = deriveEphemeralAta(vault, mint);
    const vaultAta = deriveVaultAta(mint, vault);
    const ownerAta = getAssociatedTokenAddressSync(mint, owner);
    const [shuttleEphemeralAta, shuttleBump] = deriveShuttleEphemeralAta(owner, mint, shuttleId);
    const [shuttleAta, shuttleAtaBump] = deriveShuttleAta(shuttleEphemeralAta, mint);
    const shuttleWalletAta = deriveShuttleWalletAta(mint, shuttleEphemeralAta);
    if (initVaultIfMissing) {
        instructions.push(initVaultIx(vault, mint, payer, vaultBump), initVaultAtaIx(payer, vaultAta, vault, mint), delegateEphemeralAtaIx(payer, vaultEphemeralAta, vaultEataBump, validator));
    }
    if (initAtasIfMissing) {
        instructions.push(createAssociatedTokenAccountIdempotentInstruction(payer, ownerAta, owner, mint));
    }
    if (initIfMissing) {
        instructions.push(initEphemeralAtaIx(ephemeralAta, owner, mint, payer, eataBump));
    }
    if (isPrivate) {
        instructions.push(createEataPermissionIx(ephemeralAta, payer, eataBump));
    }
    instructions.push(delegateEphemeralAtaIx(payer, ephemeralAta, eataBump, validator));
    if (amount > 0n) {
        instructions.push(setupAndDelegateShuttleEphemeralAtaWithMergeIx(payer, shuttleEphemeralAta, shuttleAta, owner, ownerAta, ownerAta, shuttleWalletAta, mint, shuttleId, shuttleBump, amount, validator));
    }
    else {
        instructions.push(initShuttleEphemeralAtaIx(payer, shuttleEphemeralAta, shuttleAta, shuttleWalletAta, owner, mint, shuttleId, shuttleBump), delegateShuttleEphemeralAtaIx(payer, shuttleEphemeralAta, shuttleAta, shuttleAtaBump, validator));
    }
    return instructions;
}
async function delegateSpl(owner, mint, amount, opts) {
    if (opts?.idempotent === false) {
        return buildDelegateSplInstructions(owner, mint, amount, opts);
    }
    return buildIdempotentDelegateSplInstructions(owner, mint, amount, opts);
}
async function delegateSplWithPrivateTransfer(owner, mint, amount, opts) {
    const payer = opts?.payer ?? owner;
    const validator = opts?.validator;
    const initIfMissing = opts?.initIfMissing ?? true;
    const initVaultIfMissing = opts?.initVaultIfMissing ?? false;
    const initAtasIfMissing = opts?.initAtasIfMissing ?? false;
    const initTransferQueueIfMissing = opts?.initTransferQueueIfMissing ?? false;
    const shuttleId = opts?.shuttleId ?? randomShuttleId();
    const minDelayMs = opts?.minDelayMs ?? 0n;
    const maxDelayMs = opts?.maxDelayMs ?? minDelayMs;
    const split = opts?.split ?? 1;
    const instructions = [];
    const [ephemeralAta, eataBump] = deriveEphemeralAta(owner, mint);
    const [vault, vaultBump] = deriveVault(mint);
    const [vaultEphemeralAta, vaultEataBump] = deriveEphemeralAta(vault, mint);
    const vaultAta = deriveVaultAta(mint, vault);
    const [queue] = (0, transferQueue_js_1.deriveTransferQueue)(mint);
    const ownerAta = getAssociatedTokenAddressSync(mint, owner);
    const [shuttleEphemeralAta, shuttleBump] = deriveShuttleEphemeralAta(owner, mint, shuttleId);
    const [shuttleAta] = deriveShuttleAta(shuttleEphemeralAta, mint);
    const shuttleWalletAta = deriveShuttleWalletAta(mint, shuttleEphemeralAta);
    if (initVaultIfMissing) {
        instructions.push(initVaultIx(vault, mint, payer, vaultBump), initVaultAtaIx(payer, vaultAta, vault, mint), delegateEphemeralAtaIx(payer, vaultEphemeralAta, vaultEataBump, validator));
    }
    if (initTransferQueueIfMissing) {
        instructions.push((0, transferQueue_js_1.initTransferQueueIx)(payer, queue, mint));
    }
    if (initAtasIfMissing) {
        instructions.push(createAssociatedTokenAccountIdempotentInstruction(payer, ownerAta, owner, mint));
    }
    if (initIfMissing) {
        instructions.push(initEphemeralAtaIx(ephemeralAta, owner, mint, payer, eataBump));
    }
    instructions.push(delegateEphemeralAtaIx(payer, ephemeralAta, eataBump, validator), depositAndDelegateShuttleEphemeralAtaWithMergeAndPrivateTransferIx(payer, shuttleEphemeralAta, shuttleAta, owner, ownerAta, ownerAta, shuttleWalletAta, mint, shuttleId, shuttleBump, amount, minDelayMs, maxDelayMs, split, validator));
    return instructions;
}
async function transferSpl(from, to, mint, amount, opts) {
    const payer = opts.payer ?? from;
    const validator = opts.validator;
    const initIfMissing = opts.initIfMissing ?? false;
    const initAtasIfMissing = opts.initAtasIfMissing ?? false;
    const initVaultIfMissing = opts.initVaultIfMissing ?? false;
    const shuttleId = opts.shuttleId ?? randomShuttleId();
    const minDelayMs = opts.privateTransfer?.minDelayMs ?? 0n;
    const maxDelayMs = opts.privateTransfer?.maxDelayMs ?? minDelayMs;
    const split = opts.privateTransfer?.split ?? 1;
    const fromAta = getAssociatedTokenAddressSync(mint, from);
    const toAta = getAssociatedTokenAddressSync(mint, to);
    if (opts.fromBalance === "ephemeral") {
        switch (opts.visibility) {
            case "private":
                if (opts.toBalance === "base") {
                    const [queue] = (0, transferQueue_js_1.deriveTransferQueue)(mint);
                    const [vault] = deriveVault(mint);
                    const vaultAta = deriveVaultAta(mint, vault);
                    return [
                        (0, transferQueue_js_1.depositAndQueueTransferIx)(queue, vault, mint, fromAta, vaultAta, toAta, from, amount, minDelayMs, maxDelayMs, split),
                    ];
                }
                if (opts.toBalance === "ephemeral") {
                    return [createTransferInstruction(fromAta, toAta, from, amount)];
                }
                break;
            case "public":
                if (opts.toBalance === "ephemeral") {
                    return [createTransferInstruction(fromAta, toAta, from, amount)];
                }
                break;
        }
    }
    const instructions = [];
    if (initVaultIfMissing) {
        const [vault, vaultBump] = deriveVault(mint);
        const [vaultEphemeralAta, vaultEataBump] = deriveEphemeralAta(vault, mint);
        const vaultAta = deriveVaultAta(mint, vault);
        instructions.push(initVaultIx(vault, mint, payer, vaultBump), initVaultAtaIx(payer, vaultAta, vault, mint), delegateEphemeralAtaIx(payer, vaultEphemeralAta, vaultEataBump, validator));
    }
    if (opts.fromBalance === "base" && initAtasIfMissing) {
        instructions.push(createAssociatedTokenAccountIdempotentInstruction(payer, fromAta, from, mint));
    }
    switch (opts.visibility) {
        case "private":
            if (opts.fromBalance === "base" && opts.toBalance === "base") {
                if (initIfMissing) {
                    instructions.push(createAssociatedTokenAccountIdempotentInstruction(payer, toAta, to, mint));
                }
                const [shuttleEphemeralAta, shuttleBump] = deriveShuttleEphemeralAta(from, mint, shuttleId);
                const [shuttleAta] = deriveShuttleAta(shuttleEphemeralAta, mint);
                const shuttleWalletAta = deriveShuttleWalletAta(mint, shuttleEphemeralAta);
                return [
                    ...instructions,
                    depositAndDelegateShuttleEphemeralAtaWithMergeAndPrivateTransferIx(payer, shuttleEphemeralAta, shuttleAta, from, fromAta, toAta, shuttleWalletAta, mint, shuttleId, shuttleBump, amount, minDelayMs, maxDelayMs, split, validator),
                ];
            }
            if (opts.fromBalance === "base" && opts.toBalance === "ephemeral") {
                if (initIfMissing) {
                    const [toEphemeralAta, toEataBump] = deriveEphemeralAta(to, mint);
                    instructions.push(createAssociatedTokenAccountIdempotentInstruction(payer, toAta, to, mint), initEphemeralAtaIx(toEphemeralAta, to, mint, payer, toEataBump), delegateEphemeralAtaIx(payer, toEphemeralAta, toEataBump, validator));
                }
                const [shuttleEphemeralAta, shuttleBump] = deriveShuttleEphemeralAta(from, mint, shuttleId);
                const [shuttleAta] = deriveShuttleAta(shuttleEphemeralAta, mint);
                const shuttleWalletAta = deriveShuttleWalletAta(mint, shuttleEphemeralAta);
                return [
                    ...instructions,
                    setupAndDelegateShuttleEphemeralAtaWithMergeIx(payer, shuttleEphemeralAta, shuttleAta, from, fromAta, toAta, shuttleWalletAta, mint, shuttleId, shuttleBump, amount, validator),
                ];
            }
            break;
        case "public":
            if (opts.fromBalance === "base" && opts.toBalance === "base") {
                return [
                    ...instructions,
                    createTransferInstruction(fromAta, toAta, from, amount),
                ];
            }
            break;
    }
    throw new Error(`transferSpl route not implemented: visibility=${opts.visibility}, fromBalance=${opts.fromBalance}, toBalance=${opts.toBalance}`);
}
async function buildIdempotentWithdrawSplInstructions(owner, mint, amount, opts) {
    const payer = opts?.payer ?? owner;
    const validator = opts?.validator;
    const initIfMissing = opts?.initIfMissing ?? true;
    const initAtasIfMissing = opts?.initAtasIfMissing ?? false;
    const shuttleId = opts?.shuttleId ?? randomShuttleId();
    const instructions = [];
    const [ephemeralAta, eataBump] = deriveEphemeralAta(owner, mint);
    const ownerAta = getAssociatedTokenAddressSync(mint, owner);
    const [shuttleEphemeralAta, shuttleBump] = deriveShuttleEphemeralAta(owner, mint, shuttleId);
    const [shuttleAta] = deriveShuttleAta(shuttleEphemeralAta, mint);
    const shuttleWalletAta = deriveShuttleWalletAta(mint, shuttleEphemeralAta);
    if (initAtasIfMissing) {
        instructions.push(createAssociatedTokenAccountIdempotentInstruction(payer, ownerAta, owner, mint));
    }
    if (initIfMissing) {
        instructions.push(initEphemeralAtaIx(ephemeralAta, owner, mint, payer, eataBump));
    }
    instructions.push(delegateEphemeralAtaIx(payer, ephemeralAta, eataBump, validator), withdrawThroughDelegatedShuttleWithMergeIx(payer, shuttleEphemeralAta, shuttleAta, owner, ownerAta, shuttleWalletAta, mint, shuttleId, shuttleBump, amount, validator));
    return instructions;
}
async function withdrawSpl(owner, mint, amount, opts) {
    if (opts?.idempotent === false) {
        const instructions = [];
        if (opts?.initAtasIfMissing === true) {
            const payer = opts.payer ?? owner;
            const ownerAta = getAssociatedTokenAddressSync(mint, owner);
            instructions.push(createAssociatedTokenAccountIdempotentInstruction(payer, ownerAta, owner, mint));
        }
        instructions.push(withdrawSplIx(owner, mint, amount));
        return instructions;
    }
    return buildIdempotentWithdrawSplInstructions(owner, mint, amount, opts);
}
function encodeAmountInstructionData(discriminator, amount, ...suffix) {
    const data = Buffer.alloc(1 + 8 + suffix.length);
    data[0] = discriminator;
    data.writeBigUInt64LE(amount, 1);
    if (suffix.length > 0) {
        data.set(suffix, 9);
    }
    return data;
}
