// @magicblock-labs/ephemeral-rollups-sdk/lib/instructions/ephemeral-spl-token-program/transferQueue.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deriveTransferQueue = deriveTransferQueue;
exports.initTransferQueueIx = initTransferQueueIx;
exports.depositAndQueueTransferIx = depositAndQueueTransferIx;
exports.ensureTransferQueueCrankIx = ensureTransferQueueCrankIx;
exports.delegateTransferQueueIx = delegateTransferQueueIx;
const web3_js_1 = require("@solana/web3.js");
const constants_js_1 = require("../../constants.js");
const pda_js_1 = require("../../pda.js");
const TRANSFER_QUEUE_SEED = Buffer.from("queue");
const TOKEN_PROGRAM_ID = new web3_js_1.PublicKey("TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA");
const INITIALIZE_TRANSFER_QUEUE_DISCRIMINATOR = 12;
const DEPOSIT_AND_QUEUE_TRANSFER_DISCRIMINATOR = 16;
const ENSURE_TRANSFER_QUEUE_CRANK_DISCRIMINATOR = 17;
const DELEGATE_TRANSFER_QUEUE_DISCRIMINATOR = 19;
function deriveTransferQueue(mint) {
    return web3_js_1.PublicKey.findProgramAddressSync([TRANSFER_QUEUE_SEED, mint.toBuffer()], constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID);
}
function initTransferQueueIx(payer, queue, mint, sizeBytes) {
    return new web3_js_1.TransactionInstruction({
        programId: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
        keys: [
            { pubkey: payer, isSigner: true, isWritable: true },
            { pubkey: queue, isSigner: false, isWritable: true },
            { pubkey: mint, isSigner: false, isWritable: false },
            { pubkey: web3_js_1.SystemProgram.programId, isSigner: false, isWritable: false },
        ],
        data: sizeBytes === undefined
            ? Buffer.from([INITIALIZE_TRANSFER_QUEUE_DISCRIMINATOR])
            : Buffer.from([
                INITIALIZE_TRANSFER_QUEUE_DISCRIMINATOR,
                ...u32le(sizeBytes),
            ]),
    });
}
function depositAndQueueTransferIx(queue, vault, mint, source, vaultAta, destination, owner, amount, minDelayMs = 0n, maxDelayMs = minDelayMs, split = 1) {
    if (!Number.isInteger(split) || split <= 0 || split > 4294967295) {
        throw new Error("split must fit in u32");
    }
    if (amount < 0n || minDelayMs < 0n || maxDelayMs < 0n) {
        throw new Error("amount and delays must be non-negative");
    }
    if (maxDelayMs < minDelayMs) {
        throw new Error("maxDelayMs must be greater than or equal to minDelayMs");
    }
    return new web3_js_1.TransactionInstruction({
        programId: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
        keys: [
            { pubkey: queue, isSigner: false, isWritable: true },
            { pubkey: vault, isSigner: false, isWritable: false },
            { pubkey: mint, isSigner: false, isWritable: false },
            { pubkey: source, isSigner: false, isWritable: true },
            { pubkey: vaultAta, isSigner: false, isWritable: true },
            { pubkey: destination, isSigner: false, isWritable: false },
            { pubkey: owner, isSigner: true, isWritable: false },
            { pubkey: TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
        ],
        data: Buffer.from([
            DEPOSIT_AND_QUEUE_TRANSFER_DISCRIMINATOR,
            ...u64le(amount),
            ...u64le(minDelayMs),
            ...u64le(maxDelayMs),
            ...u32le(split),
        ]),
    });
}
function ensureTransferQueueCrankIx(payer, queue, magicContext = constants_js_1.MAGIC_CONTEXT_ID, magicProgram = constants_js_1.MAGIC_PROGRAM_ID) {
    return new web3_js_1.TransactionInstruction({
        programId: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
        keys: [
            { pubkey: payer, isSigner: true, isWritable: true },
            { pubkey: queue, isSigner: false, isWritable: true },
            { pubkey: magicContext, isSigner: false, isWritable: true },
            { pubkey: magicProgram, isSigner: false, isWritable: false },
        ],
        data: Buffer.from([ENSURE_TRANSFER_QUEUE_CRANK_DISCRIMINATOR]),
    });
}
function delegateTransferQueueIx(queue, payer, mint) {
    return new web3_js_1.TransactionInstruction({
        programId: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
        keys: [
            { pubkey: payer, isSigner: true, isWritable: true },
            { pubkey: queue, isSigner: false, isWritable: true },
            { pubkey: mint, isSigner: false, isWritable: false },
            {
                pubkey: constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID,
                isSigner: false,
                isWritable: false,
            },
            {
                pubkey: (0, pda_js_1.delegateBufferPdaFromDelegatedAccountAndOwnerProgram)(queue, constants_js_1.EPHEMERAL_SPL_TOKEN_PROGRAM_ID),
                isSigner: false,
                isWritable: true,
            },
            {
                pubkey: (0, pda_js_1.delegationRecordPdaFromDelegatedAccount)(queue),
                isSigner: false,
                isWritable: true,
            },
            {
                pubkey: (0, pda_js_1.delegationMetadataPdaFromDelegatedAccount)(queue),
                isSigner: false,
                isWritable: true,
            },
            { pubkey: constants_js_1.DELEGATION_PROGRAM_ID, isSigner: false, isWritable: false },
            { pubkey: web3_js_1.SystemProgram.programId, isSigner: false, isWritable: false },
        ],
        data: Buffer.from([DELEGATE_TRANSFER_QUEUE_DISCRIMINATOR]),
    });
}
function u32le(n) {
    if (!Number.isInteger(n) || n < 0 || n > 4294967295) {
        throw new Error("sizeBytes out of range for u32");
    }
    return [n & 0xff, (n >>> 8) & 0xff, (n >>> 16) & 0xff, (n >>> 24) & 0xff];
}
function u64le(n) {
    if (n < 0n || n > 0xffffffffffffffffn) {
        throw new Error("value out of range for u64");
    }
    const out = Buffer.alloc(8);
    out.writeBigUInt64LE(n);
    return Array.from(out);
}
